File: 516.py
class Solution(object):
    def longestPalindromeSubseq(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) == 0:
            return 0

        matrix = []
        length = len(s)
        for i in range(length):
            matrix.append([0 for i in range(length + 1)])
        
        for i in range(len(matrix)):
            matrix[i][1] = 1

        # dp
        ret = 1
        for subLen in range(2, length + 1):
            for start in range(length - subLen + 1):
                if s[start] == s[start + subLen - 1]:
                    matrix[start][subLen] = matrix[start + 1][subLen - 2] + 2
                else:
                    matrix[start][subLen] = max(matrix[start + 1][subLen - 1], matrix[start][subLen - 1])

                if ret < matrix[start][subLen]:
                    ret = matrix[start][subLen]

        return ret

if __name__ == "__main__":
    sol = Solution()
    
    print sol.longestPalindromeSubseq("cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc")
    #print sol.longestPalindromeSubseq("bbbab")

File: 516.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
  int longestPalindromeSubseq(string s) {
	if (s.length() == 0)
	  return 0;

	int length = s.length();
    int matrix[length][length+1];

	for (int i = 0; i < length; ++i) {
	  matrix[i][0] = 0;
	  matrix[i][1] = 1;
	}

	// dp
	int ret = 1;
	for (int subLen = 2; subLen < length+1; ++subLen) {
	  for (int start = 0; start < length - subLen + 1; ++start) {
		if (s[start] == s[start + subLen - 1]) {
		  matrix[start][subLen] = matrix[start + 1][subLen - 2] + 2;
		}
		else {
		  matrix[start][subLen] = max(matrix[start + 1][subLen - 1], matrix[start][subLen - 1]);
		}

		if (ret < matrix[start][subLen])
		  ret = matrix[start][subLen];

	  }
	}

	return ret;
  }
};

int main(void) {
  Solution *sol = new Solution();
    
  cout << sol->longestPalindromeSubseq("cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc") << endl;
  cout << sol->longestPalindromeSubseq("bbbab") << endl;
}
